mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -61,136 +62,110 @@ struct RockLinalgAlignPass
   Location loc = storeOp.getLoc();
-  Value gemmOuts = storeOp.getSource();
+  Value gemmOuts = storeOp.getOperand(0);
   auto gemmOutsType = gemmOuts.getType().cast<MemRefType>();
-  int64_t sliceLength = storeOp.getLength().getSExtValue();
-  auto sliceLengthType = gemmOutsType.clone(sliceLength).cast<MemRefType>();
 
   // 1. create a second allocation of the same type to hold loaded elements
-  Value alloc = b.create<GpuAllocOp>(loc, sliceLengthType);
+  Value alloc = b.create<GpuAllocOp>(loc, gemmOutsType);
 
-  // 2. clone twcopy for <addend> -> regs as transforming_for
-  insertLoadFromOtherSource(b, loc, storeOp, inp, alloc);
-  return alloc;
-}
+  // 2. clone twcopy for <addend> into regs
+  LLVM_DEBUG(llvm::dbgs() << "Src type: " << src.getType() << " dest type: "
+                          << storeOp.getDest().getType() << "\n");
 
-Value applyTransforms(PatternRewriter &b, GlobalStoreOp gemmStoreOp, Value inp,
-                      AffineMap outToInpMap) {
-  Value ret = inp;
+  // 2.0. first move all transforms before the relocated linalg.generic
+  moveTransformsBefore(b, src);
 
-  // 1. insert broadcast op if necessary
-  MemRefType outType = gemmStoreOp.getDest().getType();
-  ret = insertTransposeAndBroadcastTransforms(b, outType.getShape(), ret,
-                                              outToInpMap);
+  // 2.1. apply transform chain from output
+  src = transformAccordingly(b, src, storeOp.getOperand(1));
 
-  // 2. also create global_store from global to regs
-  //    TODO(sjw): make sure output buffer writes (means these inputs will be
-  //    buffer reads)
-  return makeTransformingCopyLoop(b, gemmStoreOp, ret);
+  // 2.2. load into registers
+  b.create<ThreadwiseReadIntoOp>(loc, src, alloc, storeOp.getExtraViews(),
+                                 storeOp.getForceUnroll(),
+                                 storeOp.getUseIndexDiffs());
+  return alloc;
+}
+
+static Operation *traceToRealOp(Operation *op) {
+  if (auto transform = dyn_cast<rock::TransformOp>(op)) {
+    Value result = transform.getResult();
+    if (result.hasOneUse()) {
+      for (auto &use : result.getUses()) {
+        return traceToRealOp(use.getOwner());
+      }
+    }
+  }
+  return op;
 }
 
-static GlobalStoreOp traceToGlobalStore(Value inp) {
+static rock::ThreadwiseWriteAllOp traceToThreadwiseWrite(Value inp) {
   // 1. Validate that the only uses of the linalg.generic input are the one
   // generic and a copy operation or transform.
   bool allValidUses = true;
-  GlobalStoreOp result;
+  rock::ThreadwiseWriteAllOp result;
   for (Operation *use : inp.getUsers()) {
+    use = traceToRealOp(use);
     if (isa<memref::DeallocOp>(use)) {
       // ignore
       continue;
     }
-    if (isa<linalg::GenericOp>(use)) {
+    if (auto lgop = dyn_cast<linalg::GenericOp>(use)) {
       // reader
-    } else if (auto store = dyn_cast<GlobalStoreOp>(use)) {
+      if (!llvm::is_contained(lgop.inputs(), inp)) {
+        allValidUses = false;
```

Comment:
I will clean it up a bit but I prefer return-by-value (like shared_ptr) vs pass-by-reference. It is more clean from usage:
**RBV:**
`if (auto op = traceTo(use)) {
`
- uses move semantics
- MLIR ops are shared_ptr-like
- clear ownership

**PBR:**
```
Operation value;
if (succeeded(traceTo(use, value)) {
```

- not clear from call that ownership is passed
- requires more scoping
- should 'value' be tested as well, to ensure traceTo set it???

Reply:
`Operation` and friends usually get returned by value in MLIR from what I've seen.

Reply:
It is stylistic -- there are pros and cons for either style.
One could use the return channel as way of error propagation and the caller holds the ownership of the output memory (if its not smart-pointer'ish)

a middle ground might be FailureOrValue (as @krzysz00 suggested somewhere else).

but anyway it would be good to agree on a style (which I thought was more or less DPS oriented).
I think Operation is bit smart-pointy (I might be wrong -- Im saying this by just looking at existence of explicit deep copies etc) but in-case if we are returning a container it might not work well (e.g. SmallVectorImpl) -- right ?
Hence, due to asymmetry of  thing used in LLVM/MLIR, I usually preferred DPS as reference counting is not always present but like to hear what folks think...

Reply:
(Also its a nit -- feel free to ignore if you think its not applicable :) )

Reply:
There was a big discussion about this form in MLIR, lots of opinions (can't find it, maybe it was on discord). There was an Optional<> wrapper that was decidedly redundant since the Operation,Value, etc containers are all smart-ptr like. Here again, a null object return is valid.

Reply:
I am yet to find returns of llvm containers -- i.e. SmallVectorImpl -- to be smart-ptr like. They are not valid once the stack frame is destroyed (is what I find) -- (which is of course not applicable here)